extension/EdFi.Ods.Extensions.LMSX/LMS-Harmonizer/1080-lms-assignment.sql

```diff
@@ -0,0 +1,62 @@
+-- SPDX-License-Identifier: Apache-2.0
+-- Licensed to the Ed-Fi Alliance under one or more agreements.
+-- The Ed-Fi Alliance licenses this file to you under the Apache License, Version 2.0.
+-- See the LICENSE and NOTICES files in the project root for more information.
+
+CREATE OR ALTER PROCEDURE lmsx.harmonize_assignment AS
+BEGIN
+
+	INSERT INTO lmsx.[Assignment]
+		([AssignmentIdentifier]
+		,[LMSSourceSystemDescriptorId]
+		,[Title]
+		,[AssignmentCategoryDescriptorId]
+		,[AssignmentDescription]
+		,[StartDateTime]
+		,[EndDateTime]
+		,[DueDateTime]
+		,[MaxPoints]
+		,[SectionIdentifier]
+		,[LocalCourseCode]
+		,[SessionName]
+		,[SchoolYear]
+		,[SchoolId]
+		,[Id])
+	SELECT
+		lmsAssignment.AssignmentIdentifier,
+		sourceSystemDescriptor.id, -- [SourceSystem]
+		lmsAssignment.[Title],
+		assignmentCategoryDescriptor.Id, -- [AssignmentCategory]
+		lmsAssignment.[AssignmentDescription],
+		lmsAssignment.[StartDateTime],
+		lmsAssignment.[EndDateTime],
+		lmsAssignment.[DueDateTime],
+		lmsAssignment.[MaxPoints],
+		lmssection.EdFiSectionId,
+		edfiSection.LocalCourseCode,
+		edfiSection.[SessionName],
+		edfiSection.[SchoolYear],
+		edfiSection.[SchoolId],
+		lmsAssignment.SourceSystemIdentifier
+
+	FROM lms.Assignment lmsAssignment
+		INNER JOIN lms.LMSSection lmssection
+			ON lmsAssignment.LMSSectionIdentifier = lmssection.LMSSectionIdentifier
+
+		INNER JOIN edfi.Section edfiSection
+			ON lmssection.EdFiSectionId = edfiSection.Id
+
+		INNER JOIN edfi.Descriptor sourceSystemDescriptor
+			ON sourceSystemDescriptor.ShortDescription = lmsAssignment.SourceSystem
+			AND sourceSystemDescriptor.Id in (select * from lmsx.LMSSourceSystemDescriptor)
```

Comment:
I've never seen a sub-select in a join statement! I'll be curious to look at the execution plan compared to just joining to that table.

Reply:
It took me a little while to get used to this: `Id` is a Guid and I don't know why we even have it, as the joins should all be on `DescriptorId`.

```suggestion
			AND sourceSystemDescriptor.DescriptorId in (select * from lmsx.LMSSourceSystemDescriptor)
```

Reply:
My version would have this instead of the sub-selects:

```sql
		INNER JOIN edfi.Descriptor descriptor_SS
			ON descriptor_SS.CodeValue = lmsAssignment.SourceSystem
		INNER JOIN
			lmsx.LMSSourceSystemDescriptor 
		ON
			descriptor_SS.DescriptorId  = LMSSourceSystemDescriptor.LMSSourceSystemDescriptorId

		INNER JOIN edfi.Descriptor descriptor_AC
			ON descriptor_AC.CodeValue = lmsAssignment.AssignmentCategory
		INNER JOIN
			lmsx.AssignmentCategoryDescriptor
		ON
			descriptor_AC.DescriptorId = AssignmentCategoryDescriptor.AssignmentCategoryDescriptorId
```

In the end, the execution plans are structured differently but have the same performance. I have absolutely no preference between the two styles.

Reply:
Interesting! Thanks for taking the time to look at it and analyze the performance on it!

Reply:
I'll leave it with your version